### DISCUSSION TITLE
LiteLLM Proxy in Kubernetes and questions

### DISCUSSION THREAD

**Discussion by chrbrnracn** [BerriAI/litellm#1705] [General]
I'm trying to run Ollama and LiteLLM in Kubernetes. Ollama is running in a separate deployment and does not expose any external ports, only internally as a service. For now it's planned to keep Ollama separate by intention.
Now I want to run LiteLLM Proxy as OpenAI API compatible frontend for use with IDE code assisting plugins (eg. continue.dev).
* Is that possible with LiteLLM Proxy?

I'm slightly confused by the documentation.

In [Docker, Deploying LiteLLM Proxy](https://docs.litellm.ai/docs/proxy/deploy) there are examples how to deploy with database.
* A different image is used compared to the plain docker run command.
  I couldn't find any documentation on the differences. From the code it looks like litellm (Dockerfile?) pulls in some additional config file while litellm-database (Dockerfile.database?) does not by default (CMD)
* The example configures a DATABASE_URL. The example looks like an external PostgreSQL
  * What is the database used for? What is the difference to the litellm image?
  * Is a pre-deployment of a data model required or is that done implicitly by the application? Which permissions are required in the database?

"Admin UI"
The docs point to Google SSO, Microsoft SSO or a hard coded username password.
* Is there some kind of internal user management and access control (RBAC?) for additional users?
* Can other authentication mechanisms be used? eg. KeyCloak

Thanks in advance for any help.

**Comment by chrbrnracn** [comment]
Replying to myself.

> A different image is used compared to the plain docker run command.
> I couldn't find any documentation on the differences. From the code it looks like litellm (Dockerfile?) pulls in some additional config file while litellm-database (Dockerfile.database?) does not by default (CMD)

Not sure why it's implemented this way. The litellm command understands the config parameter and uses the config file when one is provided.

> The example configures a DATABASE_URL. The example looks like an external PostgreSQL
> * What is the database used for? What is the difference to the litellm image?
> * Is a pre-deployment of a data model required or is that done implicitly by the application? Which permissions are required in the database?

According to the data model I found in the code, the DB is at least used to store API keys and some user information.
Once the DB connection is established, the data model is deployed automatically.
I _**assume**_ the difference between the Docker images is, that one tries to connect to a database, the one does not. But that needs to be confirmed from someone who knows the code.

A hint for everyone who reads this after me and wants to implement it in Kubernetes. I assume you want to use Kubernetes secrets for the DB password.
There is (currently) no option to provide the password as a file. My workaround is, to set a custom entrypoint (command) which prepares and exports the DATABASE_URL variable before calling litellm.

**Comment by ishaan-jaff** [comment]
Hi @chrbrnracn thanks for this - can we hop on a call to better understand your problem ? I'm unable to repro this on my side 

Sharing my calendly for your convenience: https://calendly.com/d/4mp-gd3-k5k/litellm-1-1-onboarding-chat

**Reply by chrbrnracn** [reply]
Hi @ishaan-jaff. Thanks for the invite. It's hard to find a suitable slot due to time zone differences. Let's try to clarify the questions here. I got some more insights with my setup already, maybe others find the discussion helpful, too.
Maybe the discussion gets a bit slower, I have to work on different topics and can only proceed when I have some time left.

My setup is running locally in Kubernetes. LiteLLM is connecting to an Ollama pod and serves requests for an IntelliJ Continue.dev plugin. It's dead slow but this is another topic (no GPU available 😞). I'm running it without user management and API keys so far.

Next step for the setup is, moving this to a cloud VM where some developers shall use it. User management is the next step to take. I need to secure the setup, so no unauthorized user is able to use it.
The documentation (https://docs.litellm.ai/docs/proxy/virtual_keys) describes how to create keys for users and uses a master key for this. But I couldn't find if there's something to take care of when creating the master key, like some special team name.

**Question:** How to generate the master API key? Any special parameters required?

**Reply by ghost** [reply]
Hey @chrbrnracn the master key is set in the config 

```yaml
general_settings:
   master_key: sk-1234
```

https://docs.litellm.ai/docs/proxy/configs#quick-start

Is this what you wanted?

**Reply by ishaan-jaff** [reply]
@chrbrnracn - happy to get on a call based on your timezone 
- what email can I send an invite to ? You can DM this to me on linkedin if you prefer: https://www.linkedin.com/in/reffajnaahsi/ 
- what time works best for you?

**Reply by chrbrnracn** [reply]
@krrishdholakia I want to have an API key I can use with https://docs.litellm.ai/docs/proxy/virtual_keys

The docs say ``` # [OPTIONAL] Only use this if you to require all calls to contain this key (Authorization: Bearer sk-1234) ```
Does this mean all calls to the proxy or all calls to the configured backend?
Also I don't want _all_ calls to require this key but I want to treat this as an "admin" key which is only used to create API keys for regular users.

If this is the key I want, is it possible to define this as environment variable? I want to use Kubernetes secrets as source of this information.

**Reply by ishaan-jaff** [reply]
@chrbrnracn 
>  but I want to treat this as an "admin" key which is only used to create API keys for regular users.

That is the purpose of the master key, it's the admin key 

>  is it possible to define this as environment variable?

Yes, you can set `LITELLM_MASTER_KEY` in your env. Can you lmk if that works ?

**Reply by ishaan-jaff** [reply]
Updated our docs based on your feedback @chrbrnracn. Would love to give you a shoutout in our daily changelog what's your linkedin / twitter ? 
https://github.com/BerriAI/litellm/commit/7d670967564e7b25620cab0a2c0d22d3546873f9

**Reply by chrbrnracn** [reply]
You can use my Github handle. The LinkedIn handle looks quite ugly. :wink:

**Reply by Bruda27** [reply]
Hello!

Using VS code dev containers, I've created a devcontainer.json file that allowed Anaconda and Python 3 Environment, Im able to containerize and activate any environment like  litellm --model ollama/mistral or any open source library for enabling next generation LLM applications with multiagent, like autogen studio, of course, proxy server was enabled only to run that model, running ollama mistral doesn't require proxy configuration, so in summary, check the devcontainer.json file to ensure environment variable dependencies. 

![Screenshot 2024-02-15 195427](https://github.com/BerriAI/litellm/assets/50032691/4a7008ff-3315-4d87-b921-8afe6e59e4bf)


Im just trying to provide my POV, hope helps on something :)
